proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,209 @@
+diagnostics even without generation of debug information.
+
+> Note: The DXIL validator requires line table information to generate error
+> locations which can be inaccurate. Disabling line table information can have a
+> significant compile-time benefit.
+
+## Proposed solution
+
+### Availability Annotations
+
+In Clang, most shader APIs have header declarations that live in the default
+included `hlsl_intrinsics.h`. All shader API declarations have Clang
+availability attributes denoting version information for when the APIs are
+introduced, deprecated, and removed. This version annotation can be per-stage or
+for all stages. These annotations exist on declarations to power IDE tooling
+through clangd. Regardless of this proposal the availability annotations must be
+present.
+
+As an example of the annotations in action take the following declaration from
+`hlsl_intrinsics.h`:
+
+```c++
+__attribute__((availability(shadermodel, introduced = 6.0)))
+__attribute__((clang_builtin_alias(__builtin_hlsl_wave_active_count_bits)))
+uint WaveActiveCountBits(bool Bit);
+```
+
+> Note: the actual header uses macros to condense the attribute descriptions.
+> The example expands the macros for explicitness.
+
+The example above declares the `WaveActiveCountBits` function to require shader
+model 6.0+. This AST annotation allows clangd to communicate to users the
+availability information to language server protocol clients. This can drive
+more informed auto-complete, refactoring, and in-editor diagnostics.
+
+### Diagnostic Modes
+
+This proposal introduces three new modes for diagnosing API availability:
+relaxed, strict, and pedantic. Clang's default mode will be the
+strict mode. The `-fhlsl-diagnostic-mode=` flag allows users to select which
+diagnostic mode they wish to use.
+
+### Relaxed Diagnostic Mode
+
+The relaxed diagnostic mode performs an AST traversal after the translation unit
+has been fully parsed, and requires construction of a call graph. In the relaxed
+mode, an AST visitor will traverse to all `CallExpr` nodes that are reachable
+from exported functions (either library exports or entry functions). If the
+callee of a `CallExpr` has availability annotations that signify that the API is
+unavailable for the target shader model and stage the compiler emits a
+_warning_.
+
+Clang encodes the target shader model version in the target triple, and the
+shader stage in the `HLSLShaderAttr` which is implicitly or explicitly applied
+to the entry function.
+
+The relaxed mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions.
+
+### Strict Diagnostic Mode
+
+The implementation of the strict diagnostic mode matches the relaxed mode,
+except that when a `CallExpr` references an unavailable API, the compiler emits
+an _error_.
+
+The strict mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions.

```

Comment:
Technically the relaxed and strict modes can be implemented simply as a default-error warning, and selected explicitly with `-Werror=/-Wno-error=`. As such I'm not entirely convinced it makes sense to have a second way to do that via the `-fhlsl-diagnostic-mode` flag. I could be convinced it's worthwhile for consistency with the pedantic mode, but it's worth thinking about.